extend dagger helm chart

jholm117 opened the pull request:
I added a bunch of configuration options that my org is using and I'm hoping to get them into main so that we don't have to rely on my fork moving forward. 

I tried to make sure the values made sense for both statefulset and daemonset modes but I may have missed the mark there. Definitely open to feedback and let me know if I need to break up the PR I know there are a lot of changes.

Summary of changes:

- add support for configuring the Dagger engine via the `engine.json` file.
- Add container port and optionally bind to a host port.
- Add nodeSelector
- Add arbitrary volume mounts
- Add lifecycle preStop hook
- Add podManagementPolicy
- Add option to disable hostPath mount for run and data volumes.
- Add option to use pvc for data volume in StatefulSet

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -11,14 +11,20 @@ spec:
   # DO NOT RUN MORE THAN 1 REPLICA.
   # Only one (1) Engine is able to hold the lock, other replicas will not be able to start.
   replicas: 1
+  podManagementPolicy: {{ .Values.engine.statefulSet.podManagementPolicy }}
```
It's useful to know that this exists, but we do not expect to support more than `1` anytime soon.

jholm117 replied:
That's fair but FWIW I've been autoscaling the stateful set using EBS volumes instead of the hostpath mount to great success!

gerhard replied:
That is an interesting use-case! If you were able to make it work, do you want to capture this use-case in the comment above? Something along the lines of:

```diff
- # DO NOT RUN MORE THAN 1 REPLICA
+ # DO NOT RUN MORE THAN 1 REPLICA IF USING HOSTPATH
# Only one (1) Engine is able to hold the lock, other replicas will not be able to start.
+ # If using persistent volumes, this is not an issue. Use more than 1 replica if you know that you need it.
```

jholm117 replied:
Updated!

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -11,14 +11,20 @@ spec:
   selector:
     matchLabels:
       name: {{ include "dagger.fullname" . }}-engine
+  {{- if and .Values.engine.storage.data.persistentVolumeClaim.enabled .Values.engine.storage.persistentVolumeClaimRetentionPolicy }}
+  persistentVolumeClaimRetentionPolicy:
+    whenDeleted: {{ .Values.engine.storage.persistentVolumeClaimRetentionPolicy.whenDeleted }}
+    whenScaled: {{ .Values.engine.storage.persistentVolumeClaimRetentionPolicy.whenScaled }}
+  {{- end }}
```
Since these options are only used for the `statefulSet` I **think** that it would be best to namespace them, same as we did with `podManagementPolicy`. WDYT @jholm117 ?

jholm117 replied:
Definitely, good call.

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -80,12 +90,28 @@ spec:
               name: {{ include "dagger.fullname" . }}-magicache-token
               {{- end }}
           {{- end }}
+          {{- if .Values.engine.port }}
+          ports:
+            - name: dagger
+              containerPort: {{ .Values.engine.port }}
+              protocol: TCP
+              {{- if .Values.engine.hostPort }}
+              hostPort: {{ .Values.engine.hostPort }}
+              {{- end }}
+          {{- end }}
```
I really like how you layered the `hostPort` inside the port declaration.

Do you think that we should add a custom `--addr` arguments when this is defined @jholm117? I'm thinking:

```yaml
          args:
            - "--addr"
            - "tcp://0.0.0.0:{{ .Values.engine.port }}"
            - "--addr"
            - "unix:///var/run/buildkit/buildkitd.sock"
```

WDYT @jedevc ?

jholm117 replied:
I thought about that and decided to leave it out since I could already define it via `Values.args` and wanted to keep it simple. I think at the very least I can add a comment in the values file next to port that instructs the user to include the `--addr tcp://...` arg. Up to you!

gerhard replied:
I think that we should add these values explicitly when and the port value is defined, otherwise users get caught out when the unix port is not specified explicitly and the health checks start failing. At least that was my experience many months ago when I forgot about the need to add `--addr unix:///var/run/buildkit/buildkitd.sock` explicitly.

jholm117 replied:
I had the same experience 😂 updated!

TomChv replied:
Please look at https://github.com/dagger/dagger/pull/9866, the path will be change to `unix:///run/dagger/engine.sock` (backward compatibility is handled but might be fine to update that PR if it lands after 9866 is merged.

jholm117 replied:
Should it be `unix:///run/dagger/engine.sock` or `unix:///var/run/dagger/engine.sock`?

TomChv replied:
<img width="926" alt="Screenshot 2025-03-17 at 19 57 03" src="https://github.com/user-attachments/assets/bec912ed-b815-4273-a7ea-dc7a50fc14fb" />

I think `/var/bin/dagger/engine.sock`

jedevc replied:
Sigh, actually, no, this can't be removed because of a bug in how that engine socket is mounted.

Will open a separate PR to fix that.

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -80,12 +90,28 @@ spec:
               add:
                 - ALL
           resources: {{- toYaml .Values.engine.resources | nindent 12 }}
+          {{- if .Values.engine.lifecycle }}
+          lifecycle:
+            {{- if .Values.engine.lifecycle.preStop }}
+            preStop:
+              {{- toYaml .Values.engine.lifecycle.preStop | nindent 14 }}
+            {{- end }}
+          {{- end }}
```
Nice addition! I had a use-case a while back where I was using `dagger core engine local-cache prune` to manually free up the disk space on shutdown. FTR: https://docs.dagger.io/configuration/cache/

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -99,27 +125,63 @@ spec:
             - name: run
               mountPath: /var/run/buildkit
+            {{- end }}
             {{- if .Values.engine.config }}
             - name: config
               mountPath: /etc/dagger/engine.toml
               subPath: engine.toml
             {{- end }}
+            {{- if .Values.engine.configJson }}
+            - name: config
+              mountPath: /etc/dagger/engine.json
+              subPath: engine.json
```
What will happen if we have both of these configs @jedevc?

I am wondering whether we should only mount one of these. The fact that we are using the same name for both volume mounts - `config` - will result in an invalid manifest if both are defined. Not that likely to happen, but I can imagine the beginning of a face palm / table flip moment here.

jholm117 replied:
While there is a uniqueness constraint on volume names I don't believe there is one on volume mounts since it's not actually a name for the mount, rather a reference to the volume.

gerhard replied:
Ah yes, you are right - this is the reference! 👍

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -99,27 +125,63 @@ spec:
+            {{- end }}
       terminationGracePeriodSeconds: {{ .Values.engine.terminationGracePeriodSeconds }}
       volumes:
+        {{- if not .Values.engine.storage.data.persistentVolumeClaim.enabled }}
```
Should this be a `.Values.engine.storage.run.hostPath.enabled` condition instead? We are using a `hostPath` value which will make the connection easier to make.

jholm117 replied:
Definitely, nice catch!

gerhard commented on helm/dagger/templates/engine-statefulset.yaml:
```diff
@@ -99,27 +125,63 @@ spec:
+        {{- with .Values.engine.volumes }}
+        {{- toYaml . | nindent 8 }}
+        {{- end }}
+  {{- if .Values.engine.storage.data.persistentVolumeClaim.enabled }}
```
https://github.com/dagger/dagger/pull/9845/files#r1995602905 applies here too.

gerhard commented on helm/dagger/values.yaml:
```diff
@@ -15,6 +15,13 @@ engine:
   #   [log]
   #     format = "json"
 
+  ### Alternative: Configure Dagger Engine using JSON format
```
I think that one of these should have a higher priority. WDYT @jedevc ?

jholm117 replied:
I could be totally off-base but my understanding was they each provide unique configuration options and can coexist.

TomChv replied:
Based on https://github.com/dagger/dagger/pull/8800, `engine.json` should be prioritise since it aims to fully replace `engine.toml`  but it doesn't support by it so depending on the case, you may need to also set some options in `engine.toml` 

Based on https://github.com/dagger/dagger/blob/ecf75160867d8355b51728a0052de8bc3a1c950d/.dagger/engine.go#L105-L109 and what I could find on https://docs.dagger.io/configuration/engine/#configuration, both can cohexist with a priority over `engine.json`!

> If an option is specified in both engine.json and in engine.toml, then the engine.json settings will take precedence. 

That said, I think it's fine to have both, just reverse the 2 description so json is shown first and as the default to go for, and explain that toml should be used in addition if a specific option isn't supported by the json format.

jholm117 replied:
Sounds good to me, updated!

TomChv replied:
Thank looks good to me on this side 🚀

jedevc replied:
Yup, both can co-exist - the `engine.json` takes priority at a per-setting level.

gerhard commented on helm/dagger/values.yaml:
```diff
@@ -88,9 +110,74 @@ engine:
+      # exec:
+      #   command: ["/bin/sh", "-c", "sleep 10"]  # Example: graceful shutdown delay
+
+  ### Storage configuration for StatefulSet
```
I like that you captured the `StatefulSet` applicability @jholm117 ! WDYT about nesting this entire config under `statefulSet`?

jholm117 replied:
I moved the pvc template config under statefulSet, but I kept the hostpath config outside that scope since it applies to the daemonset as well (and renamed the keys). Let me know what you think!

jholm117 commented on helm/dagger/values.yaml:
```diff
@@ -88,9 +110,74 @@ engine:
+    # PreStop hook executed before container termination
+    preStop:
+      # exec:
+      #   command: ["/bin/sh", "-c", "sleep 10"]  # Example: graceful shutdown delay
```
> The Pod's termination grace period countdown begins before the PreStop hook is executed, so regardless of the outcome of the handler, the container will eventually terminate within the Pod's termination grace period.

I think it's realistic, maybe I'd use a longer sleep in practice. In reality I used an adaptation of this script: https://github.com/seatgeek/buildkit-prestop-script

gerhard reviewed (commented):
This looks almost there @jholm117 ! Keen to review it again once you go through the comments.

I'm happy to add the changelog fragment in case you don't get to it. [FTR](https://github.com/dagger/dagger/blob/c38d8440dc933483ca51ea7b8a54463d191844df/CONTRIBUTING.md#3-add-release-notes-fragment)

jholm117 commented:
> This looks almost there @jholm117 ! Keen to review it again once you go through the comments.
> 
> I'm happy to add the changelog fragment in case you don't get to it. [FTR](https://github.com/dagger/dagger/blob/c38d8440dc933483ca51ea7b8a54463d191844df/CONTRIBUTING.md#3-add-release-notes-fragment)

@gerhard Thanks for the review! I think I addressed all the comments and added the changelog entry but feel free to modify it!

gerhard commented:
Thanks for all the updates @jholm117 💪

Today is busier than usual, and this might spill into tomorrow, but I am keen to go over this again ASAP with the intent of having it merged in time for the next release.

cc @jedevc @matipan in case you want to take a quick look at this too, before it gets merged.

TomChv reviewed (commented):
Looks pretty good to me, left 1 important comment about `engine.json` / `engine.toml`.
I think it's mergeable after this comment is resolved.

I leave it to @gerhard for the final approval :D

jholm117 commented:
> Thanks for all the updates @jholm117 💪
> 
> Today is busier than usual, and this might spill into tomorrow, but I am keen to go over this again ASAP with the intent of having it merged in time for the next release.
> 
> cc @jedevc @matipan in case you want to take a quick look at this too, before it gets merged.

Happy to! Do I need to update the chart version?

TomChv commented:
@jholm117 I see that the `helm` CI job is failing, this need to be fixed before merging ;)

TomChv commented:
> Okay think I fixed the lint error. The check says timeout on this page, but on the Checks page it looks like it succeeded? 

Yeah looks good, I retriggered the other jobs, cgroupv2 is expected to fail until https://github.com/dagger/dagger/pull/9876 is merged, the rest should be green :D

gerhard commented on helm/dagger/templates/engine-daemonset.yaml:
```diff
@@ -54,9 +58,17 @@ spec:
+            - "--addr"
+            - "tcp://0.0.0.0:{{ .Values.engine.port }}"
+            - "--addr"
+            - "unix:///run/dagger/engine.sock"
```
I just loaded the `/run` change which came in effect since 2015: https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard . Good change 👍

gerhard reviewed (approved):
Great job @jholm117 ! This is my favourite Helm chart external contribution to date 💪

Thank you @jedevc for shepherding this and adding it all the last touches 🙏

#shipit
